pennylane/qaoa/cycle.py

```diff
@@ -75,3 +77,83 @@ def wires_to_edges(graph: nx.Graph) -> Dict[int, Tuple]:
         Dict[Tuple, int]: a mapping from wires to graph edges
     """
     return {i: edge for i, edge in enumerate(graph.edges)}
+
+
+def loss_hamiltonian(graph: nx.Graph) -> qml.Hamiltonian:
+    r"""Calculates the loss Hamiltonian for the maximum-weighted cycle problem.
+
+    We consider the problem of selecting a cycle from a graph that has the greatest product of edge
+    weights, as outlined `here <https://1qbit.com/whitepaper/arbitrage/>`__. The product of weights
+    of a subset of edges in a graph is given by
+
+    .. math:: P = \prod_{(i, j) \in E} x_{ij} c_{ij}
+
+    where :math:`E` are the edges of the graph, :math:`x_{ij}` is a binary number that selects
+    whether to include the edge :math:`(i, j)` and :math:`c_{ij}` is the corresponding edge weight.
+    Our objective is to maximimize :math:`P`, subject to selecting the :math:`x_{ij}` so that
+    our subset of edges composes a cycle.
+
+    The product of edge weights is maximized by equivalently considering
+
+    .. math:: \sum_{(i, j) \in E} x_{ij}\log c_{ij},
+
+    assuming :math:`c_{ij} > 0`.
+
+    This can be restated as a minimization of the expectation value of the following qubit
+    Hamiltonian:
+
+    .. math::
+
+        H = \sum_{(i, j) \in E} Z_{ij}\log c_{ij}.
+
+    where :math:`Z_{ij}` is a qubit Pauli-Z matrix acting upon the wire specified by the pair
+    :math:`(i, j)`. Mapping from edges to wires can be achieved using :func:`~.edges_to_wires`.
+
+    .. note::
+        The expectation value of the returned Hamiltonian :math:`H` is not equal to :math:`P`, but
+        minimizing the expectation value of :math:`H` is equivalent to maximizing :math:`P`.
+
+        Also note that the returned Hamiltonian does not impose that the selected set of edges is
+        a cycle. This constraint can be enforced using a penalty term or by selecting a QAOA
+        mixer Hamiltonian that only transitions between states that correspond to cycles.
+
+    **Example**
+
+    >>> import networkx as nx
+    >>> g = nx.complete_graph(3).to_directed()
+    >>> edge_weight_data = {edge: (i + 1) * 0.5 for i, edge in enumerate(g.edges)}
+    >>> for k, v in edge_weight_data.items():
+            g[k[0]][k[1]]["weight"] = v
+    >>>> h = loss_hamiltonian(g)
```

Comment:
Nice! Would be good to also print it.

Reply:
added here: 104ec3f3e75d83ddc2432df81a493dcff01343f0